src/tasks/UploadSlice.ts

```diff
@@ -0,0 +1,120 @@
+import { UploadResult, UploadSession } from "./LargeFileUploadTask";
+
+const binaryContentType = "application/octet-stream";
+
+/**
+ * @class
+ * Class for UploadSlice
+ */
+export class UploadSlice<T extends Parsable> {
+  constructor(
+    readonly requestAdapter: RequestAdapter,
+    readonly sessionUrl: string,
+    readonly rangeBegin: number,
+    readonly rangeEnd: number,
+    readonly totalSessionLength: number,
+    readonly parsableFactory: ParsableFactory<T>,
+    readonly errorMappings: ErrorMappings,
+  ) {}
+
+  /**
+   * Uploads a slice of the file to the server.
+   *
+   * @param {ReadableStream<Uint8Array>} stream - The stream of the file slice to be uploaded.
+   * @returns {Promise<UploadResult<T> | undefined>} - The result of the upload operation.
+   */
+  public async uploadSlice(stream: ReadableStream<Uint8Array>): Promise<UploadResult<T> | undefined> {
+    const data = await this.readSection(stream, this.rangeBegin, this.rangeEnd);
+    const requestInformation = new RequestInformation(HttpMethod.PUT, this.sessionUrl);
+    requestInformation.headers = new Headers([
+      ["Content-Range", new Set([`bytes ${this.rangeBegin}-${this.rangeEnd - 1}/${this.totalSessionLength}`])],
+      ["Content-Length", new Set([`${this.rangeEnd - this.rangeBegin}`])],
+    ]);
+    requestInformation.setStreamContent(data, binaryContentType);
+
+    const headerOptions = new HeadersInspectionOptions({ inspectResponseHeaders: true });
+    requestInformation.addRequestOptions([headerOptions]);
+
+    let itemResponse = await this.requestAdapter.send<T>(requestInformation, this.parsableFactory, this.errorMappings);
+
+    const locations = headerOptions.getResponseHeaders().get("location");
+
+    let uploadSession: UploadSession | null = null;
+    if (itemResponse) {
+      const { expirationDateTime, nextExpectedRanges } = itemResponse as Partial<UploadSession>;
+      if (nextExpectedRanges) {
+        uploadSession = {
+          expirationDateTime,
+          nextExpectedRanges: (nextExpectedRanges as unknown as string[])[0].split(","),
+        };
+        itemResponse = undefined;
+      }
+    }
+
+    return {
+      itemResponse,
+      location: locations ? (locations as unknown as string[])[0] : undefined,
+      uploadSession,
+    };
+  }
+
+  /**
+   * Reads a section of the stream from the specified start to end positions.
+   *
+   * @param {ReadableStream<Uint8Array>} stream - The stream to read from.
+   * @param {number} start - The starting byte position.
+   * @param {number} end - The ending byte position.
+   * @returns {Promise<ArrayBuffer>} - A promise that resolves to an ArrayBuffer containing the read bytes.
+   */
+  private async readSection(stream: ReadableStream<Uint8Array>, start: number, end: number): Promise<ArrayBuffer> {
+    const reader = stream.getReader();
+    let bytesRead = 0;
+    const chunks: Uint8Array[] = [];
```

Comment:
as far as I understand this implementation, we're reading all the file at once and putting it all in memory. And this every time we upload a chunk. This is going to cause significant IOPS and memory pressure.

Reply:
I've implemented a method to forward the steam to the current start position then read the stream, this will only hold the required bytes in memory. This has forced the change on the readbleStream to be a provider instead since once the chunk is read you can't reverse the stream.

Reply:
I'm not sure I understand why we can't simply read the next chunk we need here on each call? instead of reading all the chunks we might need for each call?

Reply:
Reading stops immediately when the required number  of bytes is received. There isn't a way to control the size of the chunk and the start/end position for the byte range may be part of a chunk https://github.com/microsoftgraph/msgraph-sdk-typescript-core/pull/353

Reply:
thinking about it again, it's possible to hold a chunk in memory, and start the next read request by first considering the last read chunk before continuing with the iteration. Depending on the size of the chunk, it might be a smaller memory footprint, and might be less expensive that re-initializing a new stream